### ISSUE TITLE
Replace usage of execfile with t.p.compat version

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5129]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)| @thijstriemstra reported|
|-|-|
|Trac ID|trac#5129|
|Type|task|
|Created|2011-06-01 17:42:16Z|
|Branch|https://github.com/twisted/twisted/tree/py3-execfile-5129|

From [porting-guide](http://docs.pythonsprints.com/python3_porting/py-porting.html#execfile-statement): Starting with Python 3, the `execfile` statement is no longer available. An alternative is to use the `compile()` function in conjunction with `exec()`. `compile()` can create a code object from a file, and then it can be passed into `exec()`.

```
exec(compile(source_code, source_file_name, "exec"))
```

Some of this work has been done by `loewis` in [#4244](https://github.com/twisted/twisted/issues/4244), but that ticket focuses on getting `setup.py` to run in py3k, which involves more than `execfile` compatibility.

Currently these modules use `execfile`:

```
./setup.py:                execfile(setup_py, ns, ns)
./twisted/python/test/test_release.py:        execfile("test_project", ns)
./twisted/python/test/test_release.py:        execfile("test_project", ns)
./twisted/python/_release.py:        execfile(self.directory.child("_version.py").path, namespace)
./twisted/python/dist.py:    execfile(vfile, ns)
./twisted/web/script.py:        execfile(path, glob, glob)
./twisted/web/script.py:            execfile(self.filename, namespace, namespace)
./twisted/names/authority.py:        execfile(filename, g, l)
./twisted/lore/htmlbook.py:            execfile(filename)
```

where `lore/htmlbook.py` can be ignored because its being phased out, so small enough for this single ticket.

Attachments:

* [compat3k.execfile.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1a7/1a7cb6308b16ad84931663cf2486fbfd84f7ae1a/91c81fbb098dead0185e6141970a8e81f41ceda0.diff) (5840 bytes) - added by allenap on 2011-12-07 23:08:43Z - Patch to create t.p.compat3k.execfile and use it everywhere.
* [compat3k.execfile.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1a7/1a7cb6308b16ad84931663cf2486fbfd84f7ae1a/349ab2e569cdb1410695e88b6e5b2af0749171fd.diff) (7265 bytes) - added by allenap on 2011-12-08 12:12:41Z - Patch to create t.p.compat3k.execfile and use it everywhere (with topfile and lint fixes).
* [compat3k.execfile.3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1a7/1a7cb6308b16ad84931663cf2486fbfd84f7ae1a/c3cb0d5111d190f063d4722b71f41cf833f9e0f0.diff) (6628 bytes) - added by allenap on 2012-01-11 22:54:58Z - Patch to create t.p.compat3k.execfile and use it everywhere (with review fixes).
* [compat3k.execfile.4.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1a7/1a7cb6308b16ad84931663cf2486fbfd84f7ae1a/1753ef9ff976896336a74274b29aed353c3dfdbd.diff) (11391 bytes) - added by allenap on 2012-01-21 14:28:14Z - Patch to create t.p.compat.execfile and use it everywhere (with all review fixes).
* [compat3k.execfile.5.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1a7/1a7cb6308b16ad84931663cf2486fbfd84f7ae1a/700482cfb0b394c3850b13ba146adadecf5575ec.diff) (11953 bytes) - added by allenap on 2012-04-17 22:28:42Z - Patch to create t.p.compat.execfile and use it everywhere (with universal newlines).
* [py3-execfile-5129.apply-to-r33601.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1a7/1a7cb6308b16ad84931663cf2486fbfd84f7ae1a/4b250a4c292cab9ddf1cbc66d3939d582c6f13cc.diff) (3058 bytes) - added by allenap on 2012-04-19 22:32:18Z - Like compat3k.execfile.5.diff, but can be applied to branches/py3-execfile-5129.
* [py3-execfile-5129.apply-to-r34464.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1a7/1a7cb6308b16ad84931663cf2486fbfd84f7ae1a/7e7e7a9adfee859b6c021e97b622dc5fb6d958eb.diff) (1589 bytes) - added by allenap on 2012-05-29 21:26:59Z - Tests for Java-handling code path through getExtensions, and for the "extensions" in ns branch.

<details><summary>Searchable metadata</summary>

```
trac-id__5129 5129
type__task task
reporter__thijs thijs
priority__low low
milestone__Python_3_x Python-3.x
branch__branches_py3_execfile_5129 branches/py3-execfile-5129
branch_author__exarkun__allenap exarkun, allenap
status__closed closed
resolution__fixed fixed
component__core core
keywords__py3k py3k
time__1306950136000000 1306950136000000
changetime__1342302598000000 1342302598000000
version__None None
owner__itamar itamar
cc__thijs cc__facundobatista cc__allenap cc__vperic
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="allenap's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|allenap commented|
|-|-|

I've had a go at doing this. If this looks reasonable I'm happy to figure out what I need to do to get into into a state that's good to land.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra commented|
|-|-|

Thanks. Putting it up for review as described on [ReviewProcess#Authors:Howtogetyourchangereviewed].

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [33378]) Branching to 'py3-execfile-5129'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to allenap|
|-|-|

Thanks.

  1. There's a remaining use of `__builtin__.execfile` in `twisted/lore/htmlbook.py`
  1. The `redefinition of unused 'StringIO' from line 12` warnings from pyflakes are really pyflakes bugs.  The point of pyflakes is to _improve_ your code.  I'd rather not see hacks added all over Twisted to suppress meaningless warnings.
  1. The change to `twisted/python/_release.py` to remove the nested `subprojects` import _breaks_ `DocBuilder`.
  1. The change to `inet_ntop` in `twisted/python/compat.py` seems unrelated.  Does it fix some bug?  Is there a unit test?
  1. `twisted.python.compat3k` shouldn't be a different module from `twisted.python.compat`.
  1. The tests for the new execfile implementation need docstrings and names that follow the naming convention - eg `test_execfileGlobals`.
  1. Ideally, `getExtensions` in setup.py would have unit tests as well.  It would be really awesome if you moved this function into `twisted.python.dist` and give it a unit test or two.

I did create an svn branch for this ticket, but I didn't end up putting any code in it.  Hopefully you can keep using your bzr branch and supply another patch without any trouble.

Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="allenap's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|allenap commented|
|-|-|

Thanks for the review. Sorry I've taken a while to respond; I blame
Christmas for the disruption :)

I've fixed points ![#11338](https://github.com/twisted/twisted/issues/11338) to !#6 inclusive.

In point #4 you asked:

''The change to inet_ntop in twisted/python/compat.py seems
unrelated. Does it fix some bug? Is there a unit test?''

It was a pure lint fix. I've reverted it as irrelevant to this branch.

I have not yet addressed !#7, and I know I ought to per the acceptance
criteria. I'll try and get some time for that next week.

**Comment by twisted-trac** [comment]
|<img alt="allenap's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|allenap removed owner|
|-|-|

The latest diff (the 4th) has a fix for review point #7 about getExtensions(). Please can you review it again and let me know if anything else needs doing. My fingers are crossed that it's now landable :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

is compat3k.execfile.4.diff a patch against trunk or against the branch?  It applies cleanly to both.

**Comment by twisted-trac** [comment]
|<img alt="allenap's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|allenap commented|
|-|-|

It's a diff against trunk.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_14"></a>|@thijstriemstra commented|
|-|-|

(In [33512]) apply compat3k.execfile.4.diff, refs [#5129](https://github.com/twisted/twisted/issues/5129)

**Comment by twisted-trac** [comment]
|<img alt="allenap's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|allenap commented|
|-|-|

Is there anything else I can do to help get this landed in trunk?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set owner to @exarkun|
|-|-|

  1. in `twisted/test/test_compat.py`
    1. all docstrings, even for classes, should be formatted like `"""\nfoo\n"""`
    1. There's no need to call `TestCase.setUp` from a subclass's `setUp` method.
    1. We have a nice helper for setting the contents of a file, `twisted.python.filepath.FilePath.setContent`.
    1. There's no need to delete the temporary file in `tearDown`.  It's nice to leave testing artifacts around, in case a test fails or someone wants to inspect them for some other reason.
    1. `ns_global` and `ns_local` aren't named according to the naming convention.
    1. Methods should be separated from each other by two blank lines.
  1. in `twisted/python/test/test_dist.py`
    1. Thanks for writing these :)
    1. `setup_template` is not named according to the naming convention
    1. Same comment about `setUp`
    1. Same comment about `FilePath`
    1. Same comment about two blank lines separating methods
  1. in `twisted/python/compat.py`
    1. The implementation of `execfile` here does not actually avoid triggering the Python 3 warning!  It looks like it would probably work on Python 3, but it's not clear there's any way to confirm that at the moment.  Regardless, the implementation would be simpler and would avoid the Python 3 warning if it didn't try to use the builtin `execfile` when present.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [33600]) Adjustments to meet coding standard; remove the fallback-to-builtin branch of the new execfile function: always use exec().

refs [#5129](https://github.com/twisted/twisted/issues/5129)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

(In [33601]) Construct the basedir for the getExtensions tests properly: it needs to be a package-y thing named "twisted".

refs [#5129](https://github.com/twisted/twisted/issues/5129)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set owner to allenap|
|-|-|

I fixed the points I mentioned in my review.  Buildbot points out one more problem, which is that the new implementation of `execfile` (now that it's being used) is incompatible with the builtin execfile in at least one case:

    http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/py3-execfile-5129

I haven't investigated the specifics of the incompatibility.

**Comment by twisted-trac** [comment]
|<img alt="allenap's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|allenap set owner to @exarkun|
|-|-|

`compat3k.execfile.5.diff` contains a fix for universal newline support; unrecognised line endings were confusing `compile`. It might have been easier to stick with the built-in execfile for Python 2.x ;)

exarkun, I didn't make the latest changes against your branch, for which I'm sorry. I just lost track of the conversation and didn't realise that I hadn't acted on your latest review. However, the interdiff between `compat3k.execfile.4.diff` and `compat3k.execfile.5.diff` may be of use though; there is a test in there for various newlines (though it's fairly trivial).

**Comment by twisted-trac** [comment]
|<img alt="allenap's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|allenap commented|
|-|-|

The latest patch - `py3-execfile-5129.apply-to-r33601.diff` - should apply to rev 33601 of `branches/py3-execfile-5129`, and provide the same stuff as `compat3k.execfile.5.diff` (though with updated coding style).

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

(In [34464]) Apply py3-execfile-5129.apply-to-r33601.diff 

refs [#5129](https://github.com/twisted/twisted/issues/5129)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun set owner to allenap|
|-|-|

Thanks!

  1. It looks like `ExecfileCompatTestCase.writeScript` should be passed unicode, but it's passed str more often than not.
  1. There's no test coverage for the Java-handling code path through `getExtensions`.  I think the code is fine as is, for now, but we should have a new ticket for deleting this special case or finding some way to exercise it to verify it is correct.  Will you file that ticket?
  1. There's no test coverage for the `"extensions" in ns` branch in `getExtensions`, either.  This probably should be covered now (it looks correct, fwiw).

I forced a new build with the latest code, <http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/py3-execfile-5129>.  If that turns out well, I think this will be really close to ready.  Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="allenap's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_25"></a>|allenap commented|
|-|-|

Replying to [exarkun](#note_24):
> Thanks!
> 
> 1. It looks like `ExecfileCompatTestCase.writeScript` should be
>    passed unicode, but it's passed str more often than not.

I'm not sure exactly what encoding is used when coercing a unicode to
a byte string - is it ASCII, UTF-8, sys.getdefaultencoding(), or
something else? - so I was explicit. AFAIK, .encode("ascii") is a
no-op for byte strings that are already ASCII, so this code should
work correctly in Python 2 or 3.

> 2. There's no test coverage for the Java-handling code path through
>    `getExtensions`.  I think the code is fine as is, for now, but we
>    should have a new ticket for deleting this special case or
>    finding some way to exercise it to verify it is correct.  Will
>    you file that ticket?

The attached patch - py3-execfile-5129.apply-to-r34464.diff - contains
a test for this, but I'll file a ticket to remove the special case.

> 3. There's no test coverage for the `"extensions" in ns` branch in
>    `getExtensions`, either.  This probably should be covered now (it
>    looks correct, fwiw).

That patch also contains a test for this.

> I forced a new build with the latest code, <...>. If that turns out
> well, I think this will be really close to ready. Thanks again!

My fingers are crossed :)

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_26"></a>|@vperic commented|
|-|-|

(In [34603]) Apply py3-execfile-5129.apply-to-r34464.diff

refs [#5129](https://github.com/twisted/twisted/issues/5129)

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_27"></a>|@vperic removed owner|
|-|-|

Hi allenap, you forgot to actually mark this for review, so I'm doing that now. I've also applied your patch to the branch and forced another build.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_29"></a>|@itamarst commented|
|-|-|

Looks good, I will merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_30"></a>|@itamarst set status to `closed`|
|-|-|

(In [34775]) Merge py3-execfile-5129: Switch to custom execfile() implementation, since it missing from Python 3.

Author: allenap
Reviewer: exarkun, itamar
Fixes: [#5129](https://github.com/twisted/twisted/issues/5129)
